### DISCUSSION TITLE
How do you calculate horizontally or vertically polarized gain from NF2FF results?

### DISCUSSION THREAD

**Discussion by KJ7LNW** [thliebig/openEMS-Project#164] [Q&A Usage] [answered]
xnec2c Has a feature to show the antenna gain scale either for its behavior in the horizontal avoid vertical polarization. I was looking through the [nf2ff_results](https://github.com/thliebig/openEMS/blob/master/python/openEMS/nf2ff.py#L159 ) class, but I only see circular polarization values that are extracted as follows:

```py
E_norm = 20.0*log10(nf2ff.E_norm[0]/np.max(nf2ff.E_norm[0])) + 10*log10(nf2ff.Dmax[0])
E_CPRH = 20.0*log10(np.abs(nf2ff.E_cprh[0])/np.max(nf2ff.E_norm[0])) + 10*log10(nf2ff.Dmax[0])
E_CPLH = 20.0*log10(np.abs(nf2ff.E_cplh[0])/np.max(nf2ff.E_norm[0])) + 10*log10(nf2ff.Dmax[0])
```

xnec2c [does it by adding a polarization factor](https://github.com/KJ7LNW/xnec2c/blob/master/src/draw_radiation.c#L708) to the calculated gain, however I am not familiar with the math that does this work, and I am not sure how to apply it to the openEMS result. 

If you could help point me the right direction and help me with the math, then I would appreciate it!

**Comment by VolkerMuehlhaus** [comment, marked as answer]
Looking at the code [here](https://github.com/thliebig/openEMS/blob/5f36e7f3a2367123f00999491a069aed50c6f244/python/openEMS/nf2ff.py#L174), we find

```
    * `E_theta` : theta component of electric field over frequency/theta/phi
    * `E_phi`   : phi   component of electric field over frequency/theta/phi
    * `E_norm`  : abs   component of electric field over frequency/theta/phi
    * `E_cprh`  : theta component of electric field over frequency/theta/phi
    * `E_cplh`  : theta component of electric field over frequency/theta/phi
    * `P_rad`   : radiated power (S) over frequency/theta/phi
```

The typical setup is to have the main lobe to +z direction, and "vertical" or "horizontal" is not well defined in that case. 

Instead you would look at the **theta and phi components**. The theta and phi angles in xyz coordinate system are shown in the picture below, which I borrowed from the Sonnet user manual.

If your antenna radiates to +z direction (theta=0°) the x-axis direction corresponds to phi=0°, and y-axis direction corresponds to phi=90°

![theta_phi](https://github.com/thliebig/openEMS-Project/assets/47922981/8584086b-9eda-471a-9522-101bcc59f97c)

**Reply by KJ7LNW** [reply]
oh, so do you mean that it's just as easy as changing out `nf2ff.E_cprh` for `nf2ff.E_phi` or `nf2ff.E_theta` ? eg:
```py
E_phi = 20.0*log10(np.abs(nf2ff.E_phi[0])/np.max(nf2ff.E_norm[0])) + 10*log10(nf2ff.Dmax[0])
```

**Comment by 0xCoto** [comment]
As Volker mentioned; you can use `E_theta` or `E_phi`. The only thing to note is your expression will give you the directivity and not the gain. To get the (linear) gain, multiply the (linear) directivity with the radiation efficiency:

```py
D_max = nf2ff_res.Dmax[0]
directivity_abs_linear = D_max * (nf2ff_res.E_norm[0] / np.max(nf2ff_res.E_norm[0])) ** 2

total_efficiency = nf2ff_res.Prad[0]/np.interp(farfield_frequency, frequency_range, port.P_inc)
radiation_efficiency = total_efficiency/(1 - reflection_coefficient ** 2)

gain_abs_linear = radiation_efficiency * directivity_abs_linear
```

The `reflection_coefficient` is the |S11| at the frequency of the farfield monitor. Since |S11| consists of a discrete set of points, you can interpolate the value at the `farfield_frequency`, similar to what I showed with `port.P_inc` on the preceding line. I don't use `P_acc` and `P_ref`, because I've been seeing unreasonable results (i.e., efficiency always being reported as 100%), so I compute some things manually.

To get the realized gain:

```py
realizedgain_abs_linear = total_efficiency * directivity_abs_linear
```

---

The only thing that hasn't been entirely clear to me is this: https://github.com/thliebig/openEMS-Project/discussions/147

From some observations, I believe the shapes of the patterns are correct, but the scaling seems off (the max gains don't add up correctly if you look at the results from all polarizations). I'm not sure if we have a way to compute `Dmax` for `theta`, `phi`, `cprh` and `cplh` separately, instead of using the abs (`norm`) value.

**Reply by KJ7LNW** [reply]
>  (the max gains don't add up correctly if you look at the results from all polarizations)

I've not looked at the graphs in detail, but the CP graph should be -3dB from the theta or phi graphs.  Do you know if it is?  If it is not, then I suspect there is a bug in the example or in the calculation.

**Reply by 0xCoto** [reply]
Exactly, that's more or less what I was checking and observed this issue.

**Reply by KJ7LNW** [reply]
> Exactly, that's more or less what I was checking and observed this issue.

To make sure I understood: Did you observe that phi/theta is _not_ -3dB vs CP?

If it is not -3dB:
- how far off was it?  
- what were the values?
- can you post sample code?

**Comment by VolkerMuehlhaus** [comment]
I have created a model and testcase to check the various patterns by polarization, as well as gain vs. directivity.

The model is based on the patch antenna example, but I used a different substrate (FR4) for **additional loss**, so that we have radiation efficiency of ~50% only. This gives a better view of gain vs. directivity.
You will notice that the antenna is **not matched very well** - that is on purpose, so that we can check the influence of **incident** power vs. **accepted** power in calculation of antenna efficiency. If you want to move the feed position for better return loss, see the comment in the model code.

I have also added some of the patterns for co- and cross-polarized components. The coordinate system and a view of the E field direction at 20mm above the antenna is shown for reference. 

[Test_Gain_Patch_Antenna.zip](https://github.com/thliebig/openEMS-Project/files/13692849/Test_Gain_Patch_Antenna.zip)

(All results by openEMS, except for the view field where I used another EM tool which I am more familiar with).

![lossy_patch_results](https://github.com/thliebig/openEMS-Project/assets/47922981/76bb77a5-2665-47b3-a630-52d91ed1646e)

**Reply by KJ7LNW** [reply]
Your "Direcetivity and gain" plot looks like -3dB, which I thought was correct. You are certainly more experienced than I, what is it that you see which looks like a problem that I'm missing?

I see that the xz/yz polarizations have about the same gain value, but a patch antenna that is nearly square will not have a polarization bias either way, or will it? (I see your field gradient plot, so I guess it must be polarized. . .)

I added CPRH and it looks like -3dB, too, which I would expect:

![patch antenna cprh](https://github.com/thliebig/openEMS-Project/assets/93454819/06d32001-944d-477d-ae0f-ddcd7c6b2388)

This is my patch compared to the code that you sent so you can see what I changed:

```diff
--- planar-orig.py	2023-12-16 11:52:43.505507325 -0800
+++ planar.py	2023-12-16 11:48:29.996706228 -0800
 ### Post-processing and plotting
@@ -214,23 +214,28 @@
 
     # Check directivity by polarization components
     directivity_phi   = 20.0*np.log10(nf2ff_res.E_phi[0]/np.max(nf2ff_res.E_norm[0])) + Dmax_dB
+    directivity_cprh  = 20.0*np.log10(nf2ff_res.E_cprh[0]/np.max(nf2ff_res.E_norm[0])) + Dmax_dB
     directivity_theta = 20.0*np.log10(nf2ff_res.E_theta[0]/np.max(nf2ff_res.E_norm[0])) + Dmax_dB
 
     # get field in xz plane, that is phi=0, which is index 0 
     gain_phi_xz = directivity_phi[:,0] + radiation_efficiency_dB
     gain_theta_xz = directivity_theta[:,0] + radiation_efficiency_dB
+    gain_cprh_xz = directivity_cprh[:,0] + radiation_efficiency_dB
     # get field in yz plane, that is phi=90, which is index 1 
     gain_phi_yz = directivity_phi[:,1] + radiation_efficiency_dB
     gain_theta_yz = directivity_theta[:,1] + radiation_efficiency_dB
+    gain_cprh_yz = directivity_cprh[:,1] + radiation_efficiency_dB
 
     # gain_phi in xz-plane is polarisation parallel to x axis
     # gain_phi in yz-plane is polarisation parallel to y axis
 
     figure()
-    plot(theta, np.squeeze(gain_phi_xz), 'c-', linewidth=2, label='gain (phi) xz-plane')
+    #plot(theta, np.squeeze(gain_phi_xz), 'c-', linewidth=2, label='gain (phi) xz-plane')
     plot(theta, np.squeeze(gain_phi_yz), 'b-', linewidth=2, label='gain (phi) yz-plane')
     plot(theta, np.squeeze(gain_theta_xz), 'r-', linewidth=2, label='gain (theta) xz-plane')
-    plot(theta, np.squeeze(gain_theta_yz), 'm-', linewidth=2, label='gain (theta) yz-plane')
+    #plot(theta, np.squeeze(gain_theta_yz), 'm-', linewidth=2, label='gain (theta) yz-plane')
+    plot(theta, np.squeeze(gain_cprh_xz), 'g-', linewidth=2, label='gain (CP) xz-plane')
+    plot(theta, np.squeeze(gain_cprh_yz), 'k-', linewidth=2, label='gain (CP) yz-plane')
     grid()
     ylabel('Gain (dBi)')
     xlabel('Theta (deg)')
```

**Reply by 0xCoto** [reply]
Thanks folks, I'm currently trying to debug what's wrong with my code, and this seems promising, because your values look reasonable. I've only had the chance to look at the code without being able to run it (my openEMS environment is a bit "special" at the moment). If it's not a lot of work, could you plot the following (if not, I'll try to find some time to get something working)? Let's just only focus on the XZ plane only for simplicity:

**Plot 1:**
- Directivity
- - Abs
- - Theta
- - Phi
- - cplh (LHCP)
- - cprh (RHCP)

**Plot 2:**
- Directivity vs Gain vs Realized Gain
- - Abs

If you could also report the radiation efficiency and total efficiency (%), that would help me a lot in my investigation (although I should be able to derive them from the second plot as well). Many thanks!

**Reply by VolkerMuehlhaus** [reply]
Let me start with reporting a warning that I get in the section where gain per polarization is plotted:

```
C:\Users\volker\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\matplotlib\cbook\__init__.py:1335: 
ComplexWarning: Casting complex values to real discards the imaginary part
  return np.asarray(x, float)
```

I had not noticed that before, not sure if it has any effect on results and where the imaginary part comes in. Comparing the same antenna in my commercial FDTD solver (Empire XPU), the patterns agree quite well.

**Directivity vs Gain vs Realized Gain**

Directivity vs Gain vs Realized Gain should be just fine, everything looks consistent there.

S11 = -9dB means -0.58 dB loss from reflection, that that is consistent with the number calculated from port.P_acc and port.P_inc

incident power: P_incident = 1.4984925160851587e-24 Watt
accepted power: Pin_accepted = 1.3040737182097503e-24 Watt
radiated power: Prad = 6.572829219322875e-25 Watt
directivity: Dmax = 6.64620586830729 dBi)
efficiency: nu_rad = 50.40228268955641 %, -2.975497941689813 dB
mismatch loss: 0.8702570778375781 (linear), -0.6035243589394232 dB

```
    # get accepted power into port
    Pin_accepted = np.interp(f_res, f, port.P_acc)
    # get incident power
    P_incident = np.interp(f_res, f, port.P_inc)

    # get total radiated power at nf2ff frequency (here: resonance)
    Prad_total = nf2ff_res.Prad[0]
    # get directivity (peak value) at nf2ff frequency (here: resonance)
    Dmax_dB = 10*log10(nf2ff_res.Dmax[0])
    # calculate radiation efficiency (lossless antenna is 1, regardless of input matching)
    radiation_efficiency = Prad_total/Pin_accepted
    radiation_efficiency_dB = 10.0*np.log10(radiation_efficiency)  # negative for lossy antenna
    # loss from reflection due to mismatch
    mismatch_loss =  Pin_accepted/P_incident
    mismatch_loss_dB = 10*log10(mismatch_loss)
```

I haven't plotted the circular patterns and components by polarization yet, because I don't know what exactly you want to check/compare there, at what angles and magnitudes.

Code attached
[Test_Gain_Patch_Antenna_for0xCoto.zip](https://github.com/thliebig/openEMS-Project/files/13701930/Test_Gain_Patch_Antenna_for0xCoto.zip)

**Reply by VolkerMuehlhaus** [reply]
But let me try to add this as well, here are the different polarization options and absolute value in the xz plane. 
Left is the overview, to show that Ephi is almost zero in the xz plane (sweep angle phi=0°)
Right is the zoomed view, to show that absolute value is determined by Etheta component. Both circular polarisations are 3dB down, as expected for a linearly polarized antenna.

![patch_directivity](https://github.com/thliebig/openEMS-Project/assets/47922981/a2786e4f-1361-44d3-9a4e-8bca645d5079)

This all looks correct and consistent to me.
Extended code with these additonal plots attached.

[Test_Gain_Patch_Antenna_for0xCoto.zip](https://github.com/thliebig/openEMS-Project/files/13702862/Test_Gain_Patch_Antenna_for0xCoto.zip)

**Reply by 0xCoto** [reply]
> not sure if it has any effect on results and where the imaginary part comes in

I would have imagined it describes the complex pattern (i.e. not only the directivity/gain, but also the phase), though I haven't checked. That's why I was originally trying with `np.abs()` before to get the directivity/gain. In the code, you seem to be doing that for `np.abs(nf2ff_res.E_cprh[0])` and `np.abs(nf2ff_res.E_cplh[0])`, but not `nf2ff_res.E_theta[0]` and `nf2ff_res.E_phi[0]`?

Your plots make perfect sense, so I'm trying to see where my code goes wrong. It's likely a parsing issue related to the `ComplexWarning` you mentioned, which matplotlib may handle well, but not my (somewhat more complicated) processes. Could you see what makes the warning go away while giving the same results? I tried `np.real(nf2ff_res.E_theta[0])` & `np.real(nf2ff_res.E_phi[0])`, and `np.abs(nf2ff_res.E_cprh[0])` & `np.abs(nf2ff_res.E_cplh[0])` for the circular polarizations, and while the CP results are 3 dB down (as expected), the theta and phi polarization components are almost equal to each other, which is not expected (unless my expected polarization was linear and "diagonal", which is not the case, as the Abs (`E_norm`) doesn't add up.

The efficiency calculation helped correct another issue, thanks!

**Reply by VolkerMuehlhaus** [reply]
You are right, had missed the abs() for some polarization values, and that is what the warning refers to.  Results for my plots don't change, but that might be due to the axis that I evaluated.

Updated model:
[Test_Gain_Patch_Antenna_for0xCoto.zip](https://github.com/thliebig/openEMS-Project/files/13716770/Test_Gain_Patch_Antenna_for0xCoto.zip)

Regarding your "doesn't add up" issue, the [openEMS python sources](https://github.com/thliebig/openEMS/blob/master/python/openEMS/nf2ff.py) might be interesting where the total value and circular components are calculated from theta and phi components. You can double check the difference to your code.

```
            self.E_theta.append(E_theta)
            self.E_phi  .append(E_phi)
            self.E_norm .append(np.sqrt(np.abs(E_theta)**2 + np.abs(E_phi)**2))
            self.E_cprh .append((cos_phi+1j*sin_phi) * (E_theta+1j*E_phi)/np.sqrt(2.0))
            self.E_cplh .append((cos_phi-1j*sin_phi) * (E_theta-1j*E_phi)/np.sqrt(2.0))
```

**Reply by VolkerMuehlhaus** [reply]
(deleted, I was confused which array indices belong where)

**Reply by 0xCoto** [reply]
Ok, I just checked with a commercial tool to compare the 3D patterns, and everything seems to line up perfectly. The confusion was mine and I was misinterpreting what I was looking at. Thanks for the help!

**Reply by 0xCoto** [reply]
Polarization issues solved; now I just gotta figure out why my radiation efficiency looks off (> 100%), even though the model is also a reasonable simple patch antenna...

```
S11 at frequency monitor: -4.3 dB (0.609 in linear scale)
total_efficiency: 0.6312959933534091
radiation_efficiency: 1.0072867706557003
Radiated power: Prad = 8.198164716593969e-28 W  # P_rad = nf2ff_res.Prad[0]
Incident power: Pin = 1.2986245442562965e-27 W  # P_incident = np.interp(farfield_freq, f, port.P_inc)
Accepted power: Pacc = 8.138858719704338e-28 W  # P_acc = np.interp(farfield_freq, f, port.P_acc)
```
Sometimes it's 99.9% or so, so it might be some minor floating point precision issue.

The commercial tool reports 72% and 97% for the total and the radiation efficiency, respectively. I'm not as concerned about the deviation in total efficiency, as that could be caused by slightly different material properties shifting the S11 resonance a bit (since the commercial tool is probably using frequency-dependent materials), but I'm a bit curious about the radiation efficiency.

The same radiation efficiency is obtained whether I use the formula `P_rad/P_acc` or `total_efficiency/(1 - reflection_coefficient ** 2)`.

**Reply by VolkerMuehlhaus** [reply]
> now I just gotta figure out why my radiation efficiency looks off (> 100%), even though the model is also a reasonable simple patch antenna...

That minor difference from 100% looks reasonable for a lossless antenna, so I wonder where/what losses are defined in your model?

**Reply by 0xCoto** [reply]
Indeed, after setting a lossy dielectric for the substrate, both efficiencies drop and the results don't deviate too much from the commercial tool. Thanks again, guess I'll just cap the upper bound of the efficiencies to 100% to avoid these floating point errors.

**Comment by VolkerMuehlhaus** [comment]
> Your "Direcetivity and gain" plot looks like -3dB, which I thought was correct. 

Yes, this is correct for this model which has 50% losses in the antenna dielectric, and that 50% loss causes the 3dB difference between directivity and gain in these plots. Directivity is relative to total radiated power, so that directivity is not affected by losses in the antenna.
Gain is relative to incident power, so that losses in the antenna reduce gain.

> I see that the xz/yz polarizations have about the same gain value, but a patch antenna that is nearly square will not have a polarization bias either way, or will it?

This type of patch antenna has linear polarisation, as you can see from the arrows in the E field plot at the bottom left. The E field is parallel to the x axis.

The curves are for the rotation in xz-plane (phi=0°) and rotation in the xz-plane (phi=90°).  These are also called E-plane and H plane pattern, where E-plane is the electric field direction and H-plane is perpendicular to that. 

~~

There is no circular polarization here, but now that you mentioned that: 
You are exactly right. For this linear antenna, we would see 3dB lower gain when measuring with a circular antenna to receive the signal on the other end. That is the additional -3dB that you see for the circular polarization patterns of this linear antenna.
